server/src/main/scala/org/broadinstitute/dsp/workbench/welder/server/ObjectService.scala

```diff
@@ -287,10 +287,21 @@ class ObjectService(
         IO.unit
       else IO.raiseError(InvalidLock("Fail to delocalize due to lock expiration or lock held by someone else"))
     } yield ()
+
+  private[server] def preventConcurrentAction[A](ioa: IO[A], localPath: RelativePath): IO[A] =
```

Comment:
This makes sense I think.
1. Get the Semaphore from the `Map[RelativePath, Semaphore[IO]]`
2. If the map entry is not present, create a new `Semaphore` and add it to the `Map`
3. Contruct an `IO` that `acquire`s the `Semaphore`, does the action, and `releases` the `Semaphore`

So this essentially implements a map of locks by `RelativePath`.

Reply:
yep...`acquire` on `Semaphore` will semantically block until it becomes available again. The use of `Resource` guarantees `Semaphore` will be released even if ioa fails